lib/metro_2/metro2_file.rb

```diff
@@ -24,18 +24,28 @@ def to_s
     end
 
     def trailer_from_base_segments
+      joint_segments = @base_segments.map(&:joint_segment).compact
```

Comment:
By separating out the different types of segments here, you could reduce the amount of looping below. Ideally, this would only loop through all segments once.